### DISCUSSION TITLE
How to get metadata from JSON with new let json.parse

### DISCUSSION THREAD

**Discussion by GeorgA** [savonet/liquidsoap#2812] [Q&A]
Hi,
I'm struggling with the changed JSON handling in 2.1.4 (+git@5d5c1c71) vs. 2.0. I'll need to add additional metadata (mainly cue points, start-next and fadein/out) from a database, and that worked fine in 2.0 (not showing error handling etc):

```
def handler_update_metadata_from_json(m) =
  # <get json file name from music file>
  contents=file.contents(json_file)
  metadata.json.parse(contents)    # return metadata
end
input=map_metadata(handler_update_metadata_from_json, input)
```

In 2.1.4 metadata.json.parse only returns [], don't know if that is a bug or not. I managed to dump the call altogether by simply using json.parse:

```
 contents=file.contents(json_file)
 json.parse(default=[("invalid","invalid")], contents)
```

Works, but now it complains about the deprecated json.parse() and that I should switch to "let json.parse". But I don't understand how I can get a simple key/value-list (no types, all packed into strings) like in the "old" call. I've tried this in interactive mode:

```
# line='{"stretch": 1.0}';;
json.parse(default=[("invalid","invalid")], line);;
- : [string * string] = [("stretch", "1.")]                                       ### expected result :)
- 
# let json.parse ( parsed : [(string * string)]? )=line;;
parsed : [string * string]? = null

# let json.parse ( parsed : [(string * string)] )=line;;
At line 7, char 26-45:

Error 14: Uncaught runtime error:
type: json,
message: "Parsing error: json value cannot be parsed as type [string * string]",
stack: at line 7, char 26-45,
at line 7, char 0-52
```

Is this the root problem why metadata.json.parse does no longer work or am I doing something wrong?

**Comment by vitoyucepi** [comment]
Hi @GeorgA,
This is indeed an issue.

Full reproduction
1. Generate some samples using `ffmpeg`.
   `ffmpeg -f lavfi -i "sine=frequency=1000:duration=1" -metadata artist="Artist" 1.mp3`
1. The source code of the liquidsoap script.
   ```ruby
   a = file.metadata('1.mp3')
   print(a)
   b = metadata.json.stringify(a)
   print(b)
   c = metadata.json.parse(b)
   print(c)
   ```

As for the fix.
https://github.com/savonet/liquidsoap/blob/5d5c1c716f4b7c7b2b09017b8daef03b505d5bac/libs/metadata.liq#L160-L165
I think this should look like
```ruby
def metadata.json.parse(json_string) =
  let json.parse ( metadata_list : [(string * string)] as json.object ) = json_string
  metadata_list
end
```
